File: src/bumperbot_cpp_examples/src/turtlesim_kinematics.cpp
#include <rclcpp/rclcpp.hpp>
#include <turtlesim/msg/pose.hpp>

class turtlesim_kinematics : public rclcpp::Node
{
private:
    // Define a subscription to the turtlesim pose topic
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr turtle1_pose_subscription_;
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr turtle2_pose_subscription_;
    turtlesim::msg::Pose turtle1_last_pose_;
    turtlesim::msg::Pose turtle2_last_pose_;

public:
    turtlesim_kinematics() : rclcpp::Node("turtlesim_kinematics")
    {
        // Create a subscription to the turtlesim pose topic
        RCLCPP_INFO(this->get_logger(), "turtlesim_kinematics node has been started");
        turtle1_pose_subscription_ = this->create_subscription<turtlesim::msg::Pose>(
            "turtle1/pose", 10,
            [this](const turtlesim::msg::Pose & pose)
            {
                turtle1_last_pose_ = pose;
            });

        turtle2_pose_subscription_ = this->create_subscription<turtlesim::msg::Pose>(
            "turtle2/pose", 10,
            [this](const turtlesim::msg::Pose & pose)
            {
                turtle2_last_pose_ = pose;
                // Calculate the distance between the two turtles
                
                auto translation_x = turtle2_last_pose_.x - turtle1_last_pose_.x;
                auto translation_y = turtle2_last_pose_.y - turtle1_last_pose_.y;
                auto theta = turtle2_last_pose_.theta - turtle1_last_pose_.theta;
                auto theta_degrees = theta * 180.0 / M_PI;

                RCLCPP_INFO(this->get_logger(), "Turtle translation:\n\t\
                Tx: %f\n\t\
                Ty: %f\n\
                Theta: %f degrees\n\
                Rotation Matrix:\n\t\
                |cos(Ttheta) -sin(Ttheta)|:| %f\t%f|\n\t\
                |sin(Ttheta)  cos(Ttheta)|:| %f\t%f|\n",
                        translation_x, 
                        translation_y, 
                        theta_degrees,
                        cos(theta),
                        -sin(theta),
                        sin(theta),
                        cos(theta));
            });
    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<turtlesim_kinematics>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/bumperbot_py_examples/bumperbot_py_examples/turtlesim_kinematics.py
import math
import rclpy
from rclpy.node import Node
from turtlesim.msg import Pose

class TurtleSimKinematics(Node):
    def __init__(self):
        super().__init__('turtlesim_kinematics')
        self.get_logger().info('Turtlesim Kinematics Node has been started.')
        self.turtle1_pose_sub_ = self.create_subscription(
            msg_type=Pose,
            topic='/turtle1/pose',
            callback=self.turtle1_pose_callback,
            qos_profile=10
        )
        self.turtle1_pose_sub_ = self.create_subscription(
            msg_type=Pose,
            topic='/turtle2/pose',
            callback=self.turtle2_pose_callback,
            qos_profile=10
        )
        self.last_turtle1_pose_ = Pose()
        self.last_turtle2_pose_ = Pose()

    def turtle1_pose_callback(self, msg):
        #self.get_logger().info(f'Turtle1 Pose: x={msg.x}, y={msg.y}, theta={msg.theta}')
        self.last_turtle1_pose_ = msg

    def turtle2_pose_callback(self, msg):
        #self.get_logger().info(f'Turtle2 Pose: x={msg.x}, y={msg.y}, theta={msg.theta}')
        self.last_turtle2_pose_ = msg
        Tx = self.last_turtle2_pose_.x - self.last_turtle1_pose_.x
        Ty = self.last_turtle2_pose_.y - self.last_turtle1_pose_.y

        Ttheta = self.last_turtle2_pose_.theta - self.last_turtle1_pose_.theta
        Ttheta_deg = Ttheta * 180 / math.pi
        self.get_logger().info(f'''
                               Turtle translation:\n\t
                               Tx: {Tx} \n\t
                               Ty: {Ty}\n\t
                               Ttheta: {Ttheta_deg} degrees\n
                               Rotation Matrix:\n\t
                               |cos(Ttheta) -sin(Ttheta)|:| {math.cos(Ttheta)}\t{-math.sin(Ttheta)}|\n\t
                               |sin(Ttheta)  cos(Ttheta)|:| {-math.sin(Ttheta)}\t{math.cos(Ttheta)}|\n
                               ''')

def main(args=None):
    rclpy.init(args=args)
    turtlesim_kinematics = TurtleSimKinematics()
    rclpy.spin(turtlesim_kinematics)
    turtlesim_kinematics.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()